library/std/src/io/buffered/bufwriter.rs

```diff
@@ -337,9 +336,16 @@ impl<W: Write> BufWriter<W> {
     /// ```
     #[unstable(feature = "bufwriter_into_raw_parts", issue = "80690")]
     pub fn into_raw_parts(mut self) -> (W, Result<Vec<u8>, WriterPanicked>) {
-        let buf = mem::take(&mut self.buf);
-        let buf = if !self.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
-        (self.inner.take().unwrap(), buf)
+        // SAFETY: forget(self) prevents double dropping inner/buf
+        unsafe {
+            let inner = ptr::read(&mut self.inner);
+            let buf = ptr::read(&mut self.buf);
+            let buf = if !self.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+            mem::forget(self);
```

Comment:
OT: this makes me wonder if we should have a way to destructure a `struct` that implements `Drop` without running its `drop` implementation (and maybe also a way with running it?). It feels bad having to use `unsafe` for this kind of stuff since it's simple enough that the compiler should be able to guarantee its safety in the general case.